src/hedera/hederaEngine.js

```diff
@@ -0,0 +1,512 @@
+      this.accountId = new hedera.AccountId(
+        this.walletLocalData.otherData.hederaAccount
+      )
+    }
+
+    this.earliestTimestamp = this.walletLocalData.otherData.earliestTimestamp
+    this.latestTimestamp = this.walletLocalData.otherData.latestTimestamp
+
+    if (this.accountId) {
+      this.addToLoop('getNewTransactions', 1000)
+      this.addToLoop('updateBalance', 5000)
+    }
+
+    await super.startEngine()
+  }
+
+  async resyncBlockchain(): Promise<void> {
+    await this.killEngine()
+    await this.startEngine()
+  }
+
+  async updateBalance(): Promise<void> {
+    this.log('hedera updateBalance')
+
+    if (!this.accountId) {
+      return
+    }
+
+    const balance = (await this.client.getAccountBalance(this.accountId))
+      .asTinybar()
+      .toString()
+
+    this.log('got balance:', balance)
+    this.walletLocalData.totalBalances.HBAR = balance
+    this.currencyEngineCallbacks.onBalanceChanged('HBAR', balance)
+  }
+
+  processTxs(txs: EdgeTransaction[]) {
+    if (txs) {
+      // Kabuto returns transactions in reverse chronological order
+      const latestTx = txs[0]
+      const earliestTx = txs[txs.length - 1]
+
+      if (
+        !this.latestTimestamp ||
+        latestTx.consensusAt > this.latestTimestamp
+      ) {
+        this.latestTimestamp = latestTx.consensusAt
+        this.walletLocalData.otherData.latestTimestamp = latestTx.consensusAt
+        this.walletLocalDataDirty = true
+      }
+
+      if (
+        !this.earliestTimestamp ||
+        earliestTx.consensusAt < this.earliestTimestamp
+      ) {
+        const earliestTx = txs[txs.length - 1]
+
+        this.earliestTimestamp = earliestTx.consensusAt
+        this.walletLocalData.otherData.latestTimestamp = earliestTx.consensusAt
+        this.walletLocalDataDirty = true
+      }
+
+      txs.forEach(tx => this.addTransaction(this.currencyInfo.currencyCode, tx))
+
+      this.currencyEngineCallbacks.onTransactionsChanged(txs)
+    }
+  }
+
+  async getNewTransactions() {
+    if (!this.accountId) {
+      throw new Error('no Hedera account ID')
+    }
+
+    // Kabuto returns latest transactions first
+    const txs = await this.getTransactionsKabuto({
+      after: this.latestTimestamp
+    })
+
+    if (txs) {
+      this.processTxs(txs)
+      await this.getTransactionHistory()
+    }
+  }
+
+  async getTransactionHistory() {
+    if (!this.accountId) {
+      throw new Error('no Hedera account ID')
+    }
+
+    // for robustness, we should check this again on every startup
+    // this only really matters when we're talking about testnet, which occasionally purges data
+    while (this.earliestReached) {
+      const txs = await this.getTransactionsKabuto({
+        before: this.earliestTimestamp
+      })
+
+      if (txs) {
+        this.processTxs(txs)
+      } else {
+        this.earliestReached = true
+      }
+    }
+
+    // tell the UI that we're done syncing
+    this.currencyEngineCallbacks.onAddressesChecked(1)
```

Comment:
We can't really give more granular updates right now since Kabuto only gives transactions in reverse chronological order with a hard page limit, so right now we start by fetching the latest transactions and then going backwards in time.

If it's preferable to start at the beginning and work forward, that's possible but will need to be implemented in Kabuto; however, this can then be updated in terms of `latestTx.consensusAt / Date.now()` and set to `1` when finished.

Reply:
I figured out a way to give a decent progress indicator that's basically the inverse of the above.